jina/executors/indexers/keyvalue.py

```diff
@@ -78,22 +79,17 @@ def query(self, key: int) -> Optional[bytes]:
                 return m[r:]
 
     def update(self, keys: Iterator[int], values: Iterator[bytes], *args, **kwargs):
-        # check: hard fail (raises) on key not found
-        missed = []
-        for key in keys:
-            if self.query_handler.header.get(key) is None:
-                missed.append(key)
-        if missed:
-            raise KeyError(f'Key(s) {missed} were not found in {self.save_abspath}')
-
-        # hack
-        self.query_handler.close()
-        self.handler_mutex = False
-        self.delete(keys)
+        keys = check_keys_exist(keys, self.query_handler.header.keys(), self.save_abspath, self.logger)
+        self.delete(keys, **{'keys_precomputed': True})
         self.add(keys, values)
         return
 
     def delete(self, keys: Iterator[int], *args, **kwargs):
+        if kwargs.get('keys_precomputed') is not True:
```

Comment:
why do we pass extra arguments? not expected in such key interface I would say

Reply:
the interfaces should never be changed, pasa executor parameter if needed but do not change interface or assume things will be passed

Reply:
It's a small optimization. 

When we call an `.update`, the keys will be computed for the update anyway. The update will then call the `delete` with only the filtered out keys. 

If we don't do it this way, the `.delete` method will have to recalculate and filter out the missing keys.

I can add a comment

Reply:
> the interfaces should never be changed, pasa executor parameter if needed but do not change interface or assume things will be passed

I do not EXPECT it to be there. That's why it's an optional `kwargs`. Again, just a small optimization

Reply:
in which contrxt is this optimixzation and why does it only works sometimes?

Reply:
When we call an .update, the keys will be computed for the update anyway. The update will then call the delete with only the filtered out keys.

If we don't do it this way, the .delete method will have to recalculate and filter out the missing keys

Reply:
Response to your suggestion about `_delete`

Like so ? https://github.com/jina-ai/jina/blob/3d8d678a5591861aae1608589f4a08cec7306109/jina/executors/indexers/keyvalue.py#L82

Reply:
yes more or less